File: 35.ExpressionEvaluationAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class ExpressionEvaluation {

    // Function to define precedence of operators
    static int precedence(char op) {
        if (op == '+' || op == '-') return 1;
        if (op == '*' || op == '/') return 2;
        return 0;
    }

    // Function to apply operator to two operands
    static int applyOp(int a, int b, char op) {
        if (op == '+') return a + b;
        if (op == '-') return a - b;
        if (op == '*') return a * b;
        if (op == '/') return a / b; // integer division
        return 0;
    }

    // Function to evaluate the infix expression
    static int evaluate(String expression) {
        Stack<Integer> nums = new Stack<>(); // stack for numbers
        Stack<Character> ops = new Stack<>(); // stack for operators

        int i = 0;
        while (i < expression.length()) {

            // Skip spaces
            if (expression.charAt(i) == ' ') {
                i++;
                continue;
            }

            // If current char is a digit → read full number
            if (Character.isDigit(expression.charAt(i))) {
                int val = 0;
                while (i < expression.length() && Character.isDigit(expression.charAt(i))) {
                    val = (val * 10) + (expression.charAt(i) - '0');
                    i++;
                }
                nums.push(val);
                continue;
            }

            // If '(' push it to ops
            else if (expression.charAt(i) == '(') {
                ops.push(expression.charAt(i));
            }

            // If ')' → solve until '('
            else if (expression.charAt(i) == ')') {
                while (!ops.isEmpty() && ops.peek() != '(') {
                    int b = nums.pop();
                    int a = nums.pop();
                    char op = ops.pop();
                    nums.push(applyOp(a, b, op));
                }
                ops.pop(); // remove '('
            }

            // If operator encountered
            else {
                while (!ops.isEmpty() && precedence(ops.peek()) >= precedence(expression.charAt(i))) {
                    int b = nums.pop();
                    int a = nums.pop();
                    char op = ops.pop();
                    nums.push(applyOp(a, b, op));
                }
                ops.push(expression.charAt(i));
            }
            i++;
        }

        // Apply remaining operators
        while (!ops.isEmpty()) {
            int b = nums.pop();
            int a = nums.pop();
            char op = ops.pop();
            nums.push(applyOp(a, b, op));
        }

        return nums.peek();
    }

    public static void main(String[] args) {
        String expr = "3 + 5 * ( 2 - 8 )";
        System.out.println("Expression: " + expr);
        System.out.println("Result: " + evaluate(expr));
    }
}

File: 35.ExpressionEvaluationAlgo.py
# Expression evaluation is the process of finding the result of a mathematical expression
# written in infix, postfix, or prefix form.
# Stack and queue-based algorithms help evaluate
# these expressions efficiently by following operator precedence and associativity rules.

# Explanation

# Infix expressions are the usual ones we write: 3 + 5 * 2.

# Postfix (Reverse Polish Notation): 3 5 2 * +.

# Using stacks:

# One stack for operands (numbers).

# Another stack for operators (+, -, *, /).

# Algorithm steps (for infix evaluation):

# Scan the expression left to right.

# If it’s a number → push to operand stack.

# If it’s an operator → pop operators from operator stack with higher
# or equal precedence, evaluate, then push the result back.

# If it’s ( → push to operator stack.

# If it’s ) → keep popping and evaluating until ( is found.

# After scanning, evaluate remaining operators.

def precedence(op):
    if op == '+' or op == '-':
        return 1
    if op == '*' or op == '/':
        return 2
    return 0


def apply_op(a, b, op):
    if op == '+':
        return a + b
    if op == '-':
        return a - b
    if op == '*':
        return a * b
    if op == '/':
        return a // b


def evaluate(expression):
    nums = []
    ops = []
    i = 0
    while i < len(expression):
        if expression[i] == ' ':
            i += 1
            continue

        elif expression[i].isdigit():
            val = 0
            while i < len(expression) and expression[i].isdigit():
                val = (val * 10) + int(expression[i])
                i += 1
            nums.append(val)
            continue

        elif expression[i] == '(':
            ops.append(expression[i])

        elif expression[i] == ')':
            while ops and ops[-1] != '(':
                b = nums.pop()
                a = nums.pop()
                op = ops.pop()
                nums.append(apply_op(a, b, op))
            ops.pop()

        else:
            while ops and precedence(ops[-1]) >= precedence(expression[i]):
                b = nums.pop()
                a = nums.pop()
                op = ops.pop()
                nums.append(apply_op(a, b, op))
            ops.append(expression[i])
        i += 1

    while ops:
        b = nums.pop()
        a = nums.pop()
        op = ops.pop()
        nums.append(apply_op(a, b, op))

    return nums[-1]


expr = "3 + 5 * ( 2 - 8 )"
print("Expression:", expr)
print("Result:", evaluate(expr))
